nvtabular/workflow.py

```diff
@@ -734,83 +718,92 @@ def apply(
         if not self.phases:
             self.finalize()
+
+        # Gather statstics (if apply_offline), and/or transform
+        # and write out processed data
         if apply_offline:
-            if hugectr_gen_output:
-                raise ValueError(
-                    "TODO: Support HugeCTR output for offline processing with Dask."
-                    " This is part of the larger task of aligning online/offline API."
-                )
-            self.update_stats(
+            self.build_and_process_graph(
                 dataset,
                 output_path=output_path,
                 record_stats=record_stats,
                 shuffle=shuffle,
+                output_format=output_format,
                 out_files_per_proc=out_files_per_proc,
             )
         else:
-            shuffler = None
-            huge_ctr = None
-            if shuffle:
-                if isinstance(shuffle, str):
-                    raise ValueError("TODO: Align shuffling/writing API for online/offline.")
-                shuffler = nvt_io.Shuffler(output_path, num_out_files=out_files_per_proc)
-            if hugectr_gen_output:
-                self.cal_col_names = False
-                if hugectr_output_format == "binary":
-                    huge_ctr = nvt_io.HugeCTRWriter(
-                        hugectr_output_path, num_out_files=hugectr_num_out_files
-                    )
-                elif hugectr_output_format == "parquet":
-                    huge_ctr = nvt_io.ParquetWriter(
-                        hugectr_output_path, num_out_files=hugectr_num_out_files
-                    )
-            # "Online" apply currently requires manual iteration
+            self.iterate_online(
+                dataset,
+                output_path=output_path,
+                shuffle=shuffle,
+                output_format=output_format,
+                out_files_per_proc=out_files_per_proc,
+            )
+
+    def iterate_online(
+        self,
+        dataset,
+        end_phase=None,
+        output_path=None,
+        shuffle=None,
+        output_format=None,
+        out_files_per_proc=None,
+        apply_ops=True,
+    ):
+        """ Iterate through dataset and (optionally) apply/shuffle/write.
+        """
+        # Check if we have a (supported) writer
+        writer = None
+        writer_cls = None
+        if output_format:
+            if output_format == "parquet":
+                writer_cls = nvt_io.ParquetWriter
+            elif output_format == "hugectr":
+                writer_cls = nvt_io.HugeCTRWriter
+            else:
+                raise ValueError("Output format not yet supported.")
+            output_path = output_path or "./"
+            output_path = str(output_path)
+            fs = get_fs_token_paths(output_path)[0]
+            writer = writer_cls(
+                output_path, num_out_files=out_files_per_proc, shuffle=shuffle, fs=fs
+            )
```

Comment:
Can we move this block of code into a factory function in nvtabular/io.py ?  Something like a function that takes the output_format and output_path and returns a Writer object.

Reply:
Good suggestion - I added a `writer_factory` function in `io.py`.  Let me know what you think